### ISSUE TITLE
Support Ctrl+C in TrainEndCheckpoint

### ISSUE THREAD

**Issue by thomasjpfan** [dnouri/skorch#387]
Continuing from PR https://github.com/dnouri/skorch/pull/360

When one presses Ctrl+C in the middle of an epoch, the latest epoch would most likely be in an incomplete state. When the checkpoint is reloaded and training starts again, I propose throwing away the incomplete epoch and start training on that epoch.

For example, if one Ctrl+C's during epoch 9, only the history up to epoch 9, i.e. 1-8, will be saved. All the batch data in epoch 9 would be discarded.

**Comment by BenjaminBossan** [comment]
It makes sense to be aware of an epoch being incomplete. Do you have a proposal how exactly to implement this change? E.g., is it confined to `TrainEndCheckpoint` or on the `NeuralNet`? 

We should also think about the use case of ctrl+c. For me, it's more about doing some quick experimentation but it also used in some functions to interrupt training.

**Comment by thomasjpfan** [comment]
If this was implemented, we could add a `on_exception` callback function, and have `NeutralNet` catch the exception and pass it on to the callbacks. before fully exiting.

One possible use case for `ctrl-c` is as follows:

0. Start training model.
1. Validation loss is not decreasing anymore.
2. Training is stopped.
3. `TrainEndCheckpoint` saves the most recent checkpoint.
4. Learning rate is lowered.
5. Training resumes from the most recent checkpoint with the lower learning rate.

**Comment by BenjaminBossan** [comment]
How would you go about throwing away the incomplete data from the interrupted epoch?

**Comment by thomasjpfan** [comment]
In the `on_exception` callback, the history would have be trimmed down. The drawback with this approach is that the module will not be in sync with the history. The module would be trained a little more than the history is stating.

**Comment by BenjaminBossan** [comment]
> The drawback with this approach is that the module will not be in sync with the history

This is also my fear.

Additionally, when a callback trims the history, this could be an unforeseen side-effect to some, the source of which is hard to track down. Maybe there is a solution that is more explicit, like calling a trim method directly on the history.

**Comment by benjamin-work** [comment]
> We can save the the partial history and continue training from that point ... This would lead to an epoch with 15 iterations while other epochs would have 10 iterations.

Maybe I misunderstand but isn't this what happens at the moment?

Say we have the case from your example. I wonder if it's possible to take note of the number of iterations (5) and when we `partial_fit`, start from the next iteration (6), so that all epochs have exactly the right number of iterations (10). But that idea would also need further investigation.

Regardless of that, would it be helpful to have an explicit `trim` method on `history`?

**Comment by thomasjpfan** [comment]
> Maybe I misunderstand but isn't this what happens at the moment?

Currently, there is no callback that saves a history in the middle of an epoch. All epochs will have the same number of training iterations.

> would it be helpful to have an explicit trim method on history?

If we adopt the first proposal, then a `trim` method on `history` would be helpful.

> when we partial_fit, start from the next iteration (6), so that all epochs have exactly the right number of iterations (10). 

This would involve changing the training loop to have it skip some iterations. It may not be worth it to complicate the training loop just to support this feature.

**Comment by benjamin-work** [comment]
Okay, so also I believe we should keep it as simple as possible. When training is interrupted, there will some dangling batches, it's not like we can roll this back.

What is the simplest way to ensure those batches are not lost from history because it was not saved? We could add an `on_exception` method. I wonder if a simpler way is possible (say trigger `on_train_end` after a keyboard interrupt?). Then again, this method could be useful for other use cases. Any opinions?

**Comment by ottonemo** [comment]
> Okay, so also I believe we should keep it as simple as possible. When training is interrupted, there will some dangling batches, it's not like we can roll this back.
> 
> What is the simplest way to ensure those batches are not lost from history because it was not saved? We could add an `on_exception` method. I wonder if a simpler way is possible (say trigger `on_train_end` after a keyboard interrupt?). Then again, this method could be useful for other use cases. Any opinions?

In case of `^C` the `on_train_end` callback is executed anyway, so any non-error condition is handled gracefully already and it is the obligation of the checkpoint callback to handle dangling history values in that case.

Generally I agree that we should not roll back anything from the history and if there are inconsistent number of batches per epoch so be it. We just have to make sure that no code depends on a fixed number of batches per epoch.

**Comment by thomasjpfan** [comment]
Reviewing the code, It looks like this feature is already supported. Since `KeyboardInterrupt` is already caught, and `on_train_end` is triggered, saving the checkpoint.

> We just have to make sure that no code depends on a fixed number of batches per epoch.

This would be where most of the work lies. There are pieces of code the assumes that `len(history)` is equal the to the number of completed epochs.

**Comment by benjamin-work** [comment]
Do you have a proposal how to measure the real number of completed epochs?

**Comment by BenjaminBossan** [comment]
I think the idea is sound but I would not rely on an indirect signal such as `dur`. Perhaps we should add a column to `history` during `NeuralNet.on_epoch_end` that does nothing more but indicate that the epoch was completed regularly. (This column should be ignored by `PrintLog`.)

**Comment by thomasjpfan** [comment]
This is already supported by `TrainEndCheckpoint`. When training is continued after a Ctrl-C event, a new epoch is started, and the incomplete epoch is left in the history.
